apps/04TowersOfHanoi.js

```diff
@@ -1,93 +1,129 @@
 'use strict';
-
+// the following is for node input:
 var assert = require('assert');
 var readline = require('readline');
 var rl = readline.createInterface({
-  input: process.stdin,
-  output: process.stdout
+    input: process.stdin,
+    output: process.stdout
 });
-
+// board is here so it won't change every time:
 var stacks = {
-  a: [4, 3, 2, 1],
-  b: [],
-  c: []
+    a: [4, 3, 2, 1],
+    b: [],
+    c: []
 };
 
 function printStacks() {
-  console.log("a: " + stacks.a);
-  console.log("b: " + stacks.b);
-  console.log("c: " + stacks.c);
+    console.log("a: " + stacks.a);
+    console.log("b: " + stacks.b);
+    console.log("c: " + stacks.c);
 }
 
-function movePiece() {
-  // Your code here
-
+function movePiece(startStack, endStack) {
+    // Barbara! Fix this - comment it out and move to Hanoi funciton.
+    // Your code here
+    // remove the block;
+    // var startMove =
+      stacks[endStack].push(stacks[startStack].pop());
+    // save the block:
+    // var block = startMove;
+    // put the block on the selected tower:
+    // var endMove =
+  ;
 }
 
-function isLegal() {
-  // Your code here
-
+function isLegal(startStack, endStack) {
+    // Your code here
+    // var startMove = stacks[startStack].pop();
+    // var endMove = stacks[endStack].push();
+    return !(stacks[startStack][stacks[startStack].length - 1] >
+        stacks[endStack][stacks[endStack].length - 1]);
 }
 
 function checkForWin() {
-  // Your code here
-
+    // Your code here
+    var win = stacks.b.length === 4 || stacks.c.length === 4;
+    if (win) {
+        console.log("You Won!");
+        // process.exit();
+    }
+    return win;
 }
 
 function towersOfHanoi(startStack, endStack) {
-  // Your code here
-
+    // Your code here
+    // Is legal
+    // Move piece
+    // check for win
+
+    isLegal(startStack, endStack);
+    movePiece(startStack, endStack);
```

Comment:
Danger! Danger!

You execute movePiece() on line 60 regardless of the outcome of isLegal(). This will allow illegal moves. You need to bind these functions together with an `if(){ ...}` statement.